docs/design/unsupported-args-design.md

```diff
@@ -0,0 +1,118 @@
+# Unsupported Arguments schema and common design
+Date: 2024-06-29
+
+## Abstract
+OADP allows to configure it's components parameters via number of DataProtectionApplication Schema.
+
+Unsupported arguments within the OADP operator are those that do not conform to the standard DataProtectionApplication Schema or are explicitly identified as unsupported or custom arguments.
+
+There is need to provide a common way for defining and managing unsupported arguments within the OADP operator for its' components such as velero server or node-agent arguments.
+
+## Background
+The OADP allows configuration of several predefined arguments via the DataProtectionApplication Schema. These parameters are carefully selected and tested to meet the highest quality standards. However, in some circumstances, there is a need to pass other parameters that may be under development, not yet fully tested, or specifically overriding the default ones.
+
+Over the past OADP releases, some implementations have provided access to extra arguments. However, there has been no standardization, and unknown to DataProtectionApplication Schema arguments were not allowed.
+
+## Goals
+- Define common way to allow passing extra arguments to various executables used by OADP.
+- Ensure the new common way do not interfere with already existing implementations.
+
+## Non Goals
+- Deprecate current ways of defining extra arguments.
+- Change precedences within current ways of defining extra arguments.
+- Validation of unsupported args, they are passed to the appropriate executables without any prior validation.
+
+## Design
+Unsupported arguments are defined as a `ConfigMap` within the same Namespace as OADP Operator.
+
+There may be multiple `ConfigMaps` for particular section of the OADP configuration.
+
+There are default naming convention for the `ConfigMaps`, however user may override those using `DPA` annotations.
+
+When ConfigMap is found, controller takes those values as the highest priority and replaces **all** of the other configuration options, even if they were defined within `DPA` schema for the section for which the `ConfigMap` was created.
+
+The Schema of an `ConfigMap` includes data that corresponds to the arguments passed to the relevant executable as in the below example:
+
+   ```
+   apiVersion: v1
+   kind: ConfigMap
+   metadata:
+     name: oadp-unsupported-velero-server-args
+     namespace: openshift-adp
+   data:
+     --default-volume-snapshot-locations: 'aws:backups-primary,azure:backups-secondary'
+     --log-level: 'debug'
```

Comment:
one fear with this approach is if there is any flag that we just activate by passing it. But if that happens, we can just pass an empty string here and in the logic to passing flags, just check value length, right?

Reply:
@mateusoliveira43 
You are very much right here, we may want to have empty string to represent boolean value, for example:
```yaml
  --enable-some-option
```

But `ConfigMap` does not allow empty strings, so I wonder if we want to either include some constant string that represents the empty string or boolean flag? Can't think of other option?

Reply:
Empty string like this, is accepted, right? i.e. a string of length zero `''`
```yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: mateus-test
  namespace: openshift-adp
  uid: e05e4bde-e250-4df6-a90c-8c29bbe7a07d
  resourceVersion: '380824521'
  creationTimestamp: '2024-06-06T13:11:38Z'
  managedFields:
    - manager: Mozilla
      operation: Update
      apiVersion: v1
      time: '2024-06-06T13:11:38Z'
      fieldsType: FieldsV1
      fieldsV1:
        'f:data':
          .: {}
          'f:--another': {}
          'f:--log-level': {}
data:
  '--another': ''
  '--log-level': ''
```

From what I searched, that would not affect us
```
$ go run cmd/velero/velero.go help server | grep empty
      --log_dir string                   If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                  If non-empty, use this log file (no effect when -logtostderr=true)
```

But constant string that represents the empty string seems good idea

Reply:
Can we just use "true" here? Bool arg without value is just a short-cut for true. If we require a constant value anyway, lets just have the users explicitly set it to true. We can just document that for bool args, an explicit true or false is needed.

Reply:
I think that would work, from what I looked, only fields like this are boolean

Reply:
I believe that in some CLI's it may not work, so passing unexpected arg (empty string) may actually cause error ?

Reply:
@mpryc My suggestion was just to require args. This is only relevant for bool args where "=true" can be left off with no change in meaning. Via the configmap, lets just require them to specify `true` explicitly -- that way we hve no unexpected empty strings.

Reply:
@sseago that will it work, so for the arguments that are boolean we require `=true` or `=false` to be passed, however we don't validate those.

Below is a full list of server args from `velero server` and in our case for example `--default-snapshot-move-data` would require `--default-snapshot-move-data=true` right?

```shell
velero server --help
Run the velero server

Usage:
  velero server [flags]

Flags:
      --backup-sync-period duration                         How often to ensure all Velero backups in object storage exist as Backup API objects in the cluster. This is the default sync period if none is explicitly specified for a backup storage location. (default 1m0s)
      --client-burst int                                    Maximum number of requests by the server to the Kubernetes API in a short period of time. (default 30)
      --client-page-size int                                Page size of requests by the server to the Kubernetes API when listing objects during a backup. Set to 0 to disable paging. (default 500)
      --client-qps float32                                  Maximum number of requests per second by the server to the Kubernetes API once the burst limit has been reached. (default 20)
      --default-backup-storage-location string              Name of the default backup storage location. DEPRECATED: this flag will be removed in v2.0. Use "velero backup-location set --default" instead. (default "default")
      --default-backup-ttl duration                         How long to wait by default before backups can be garbage collected. (default 720h0m0s)
      --default-item-operation-timeout duration             How long to wait on asynchronous BackupItemActions and RestoreItemActions to complete before timing out. Default is 4 hours (default 4h0m0s)
      --default-repo-maintain-frequency duration            How often 'maintain' is run for backup repositories by default.
      --default-snapshot-move-data                          Move data by default for all snapshots supporting data movement.
      --default-volume-snapshot-locations mapStringString   List of unique volume providers and default volume snapshot location (provider1:location-01,provider2:location-02,...)
      --default-volumes-to-fs-backup                        Backup all volumes with pod volume file system backup by default.
      --disable-controllers strings                         List of controllers to disable on startup. Valid values are backup,backup-operations,backup-deletion,backup-finalizer,backup-sync,download-request,gc,backup-repo,restore,restore-operations,schedule,server-status-request
      --disable-informer-cache                              Disable informer cache for Get calls on restore. WIth this enabled, it will speed up restore in cases where there are backup resources which already exist in the cluster, but for very large clusters this will increase velero memory usage. Default is false (don't disable).
      --fs-backup-timeout duration                          How long pod volume file system backups/restores should be allowed to run before timing out. (default 4h0m0s)
      --garbage-collection-frequency duration               How often garbage collection is run for expired backups.
  -h, --help                                                help for server
      --item-operation-sync-frequency duration              How often to check status on backup/restore operations after backup/restore processing. Default is 10 seconds
      --log-format                                          The format for log output. Valid values are text, json. (default text)
      --log-level                                           The level at which to log. Valid values are trace, debug, info, warning, error, fatal, panic. (default info)
      --max-concurrent-k8s-connections int                  Max concurrent connections number that Velero can create with kube-apiserver. Default is 30. (default 30)
      --metrics-address string                              The address to expose prometheus metrics (default ":8085")
      --plugin-dir string                                   Directory containing Velero plugins (default "/plugins")
      --profiler-address string                             The address to expose the pprof profiler. (default "localhost:6060")
      --resource-timeout duration                           How long to wait for resource processes which are not covered by other specific timeout parameters. Default is 10 minutes. (default 10m0s)
      --restore-only                                        Run in a mode where only restores are allowed; backups, schedules, and garbage-collection are all disabled. DEPRECATED: this flag will be removed in v2.0. Use read-only backup storage locations instead.
      --restore-resource-priorities stringArray             Desired order of resource restores, the priority list contains two parts which are split by "-" element. The resources before "-" element are restored first as high priorities, the resources after "-" element are restored last as low priorities, and any resource not in the list will be restored alphabetically between the high and low priorities. (default customresourcedefinitions,namespaces,storageclasses,volumesnapshotclass.snapshot.storage.k8s.io,volumesnapshotcontents.snapshot.storage.k8s.io,volumesnapshots.snapshot.storage.k8s.io,datauploads.velero.io,persistentvolumes,persistentvolumeclaims,serviceaccounts,secrets,configmaps,limitranges,pods,replicasets.apps,clusterclasses.cluster.x-k8s.io,endpoints,services,-,clusterbootstraps.run.tanzu.vmware.com,clusters.cluster.x-k8s.io,clusterresourcesets.addons.cluster.x-k8s.io)
      --store-validation-frequency 0s                       How often to verify if the storage is valid. Optional. Set this to 0s to disable sync. Default 1 minute. (default 1m0s)
      --terminating-resource-timeout duration               How long to wait on persistent volumes and namespaces to terminate during a restore before timing out. (default 10m0s)
      --uploader-type string                                Type of uploader to handle the transfer of data of pod volumes (default "restic")

Global Flags:
      --add_dir_header                   If true, adds the file directory to the header of the log messages
      --alsologtostderr                  log to standard error as well as files (no effect when -logtostderr=true)
      --colorized optionalBool           Show colored output in TTY. Overrides 'colorized' value from $HOME/.config/velero/config.json if present. Enabled by default
      --features stringArray             Comma-separated list of features to enable for this Velero process. Combines with values from $HOME/.config/velero/config.json if present
      --kubeconfig string                Path to the kubeconfig file to use to talk to the Kubernetes apiserver. If unset, try the environment variable KUBECONFIG, as well as in-cluster configuration
      --kubecontext string               The context to use to talk to the Kubernetes apiserver. If unset defaults to whatever your current-context is (kubectl config current-context)
      --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                   If non-empty, write log files in this directory (no effect when -logtostderr=true)
      --log_file string                  If non-empty, use this log file (no effect when -logtostderr=true)
      --log_file_max_size uint           Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                      log to standard error instead of files (default true)
  -n, --namespace string                 The namespace in which Velero should operate (default "openshift-adp")
      --one_output                       If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
      --s2a_enable_appengine_dialer      If true, opportunistically use AppEngine-specific dialer to call S2A.
      --s2a_timeout duration             Timeout enforced on the connection to the S2A service for handshake. (default 3s)
      --skip_headers                     If true, avoid header prefixes in the log messages
      --skip_log_headers                 If true, avoid headers when opening log files (no effect when -logtostderr=true)
      --stderrthreshold severity         logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
  -v, --v Level                          number for the log level verbosity
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
```

Reply:
@sseago how would you go about this:

```
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: oadp-unsupported-velero-server-args
    namespace: openshift-adp
  data:
    --default-volume-snapshot-locations: 'aws:backups-primary,azure:backups-secondary'
    --log-level: 'debug'
    --default-snapshot-move-data: '=true'
```

We can see that some arguments must have space between value and for `boolean` there must be no space, which isn't that clear from the above ConfigMap example.

e.g. how to distinguish between:

`velero --my-sample-arg=true --my-sample-arg true`

Reply:
@sseago please review current change to the design, it uses `=true` or `=false` to indicate boolean args, so they need to start with `=` as a first value character.

Reply:
Because code speak itself, there is a small function that implements ConfigMap to args on how this design describes it:

https://github.com/openshift/oadp-operator/pull/1423

In our case it's going to replace this in the Deployment:

https://github.com/vmware-tanzu/velero/blob/main/pkg/install/deployment.go#L290

Reply:
I think there is no need for `=true`, we just do logic with string being `"%s=%s", key, value`, for all flags

Reply:
If the operator-manager ( I think ) is too strict in how it reformats commands and enforces too much we run the risk of making a mistake or limiting the future proofness off this design.  
An Example:
1. If a customer typos a config value in the configmap, and there is an error in velero server args or node-agent the velero pod will not come up and there will be an error.  In this case the customer ( NOTE: along side engineering) will have to fix the configmap.
2.  If the code that transforms the configmap has a mistake or error or didn't account for something engineering will have to ship the customer new code.

Noting that this feature is undocumented and "should" only be employed in partnership with engineering I would much rather have to deal with #1 vs #2

Garbage in -->  Garbage out is a MUCH easier situation to deal with than updating the controller-manager code. In fact we do not have a way to ship a custom controller-manager.  
Being too strict with how arguments are formated has the potential to defeat the whole purpose of the design.

Reply:
@mpryc I think we can make a couple simplifications:
1) We can skip the "--" in the string the user is adding. `--foo` -- the name is "foo", the "--" is just how flags are passed
2) `--param-name=param-value` should work in all cases, no need for "=" in some names, and no need to worry about spacing.

Basically, you have k/v pairs in the config map, and then when generating the flags, `"--" + paramName + "=" + paramValue`

So the above configmap would look like this:
```
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: oadp-unsupported-velero-server-args
    namespace: openshift-adp
  data:
    default-volume-snapshot-locations: 'aws:backups-primary,azure:backups-secondary'
    log-level: 'debug'
    default-snapshot-move-data: 'true'
```

Reply:
+1 @sseago

Reply:
Ok so I have aligned with all the comments and updated design doc with function that implements it available for review at https://github.com/openshift/oadp-operator/pull/1423

The above function is commented with numerous test scenarios to present how this behaves with different ConfigMap value/key pairs. Please review and let me know if current design / implementation can be final, so I will integrate with oadp.

@sseago @shubham-pampattiwar @mateusoliveira43 ^^